### PR TITLE
[NTGDI][GDI32_APITEST] Allow calling NtGdiCreateHalftonePalette with HDC=NULL

### PR DESCRIPTION
**Pull request by xmine64**
## Purpose

Currently ReactOS doesn't allow calling `CreateHalftonePalette(HDC)` with NULL, while tests show this is allowed on Windows and even Wine. Also, SDK headers provided by Microsoft [here](https://github.com/microsoft/win32metadata/blob/00506a453b49a9a6e8947f69c00ae1c1ec4cbb41/generation/WinSDK/RecompiledIdlHeaders/um/wingdi.h#L5117) indicates that `CreateHalftonePalette`'s parameter is optional, and can be NULL.

This behavior causes compatibility issues, breaking applications like *Internet Explorer 8.0* and *Microsoft Encarta*.

This PR imports a patch provided by simonelombardo in Jira which allows `NtGdiCreateHalftonePalette` to run with NULL HDC, and also adds APITEST for CreateHalftonePalette.

The test result without the patch:

image

The test result after applying the patch:

image

Test result on other platforms:

image

image

JIRA issue: [CORE-20231](https://jira.reactos.org/browse/CORE-20231)

## Proposed changes

* Add APITEST for CreateHalftonePalette: The test calls `CreateHalftonePalette(NULL)` and verifies it returns a valid palette
* Make `NtGdiCreateHalftonePalette` accept NULL: Don't fail when HDC was NULL and only lock it when it's not

## Testbot runs (Filled in by Devs)

- [x] KVM x86: https://reactos.org/testman/compare.php?ids=108741,108754
- [x] KVM x64: https://reactos.org/testman/compare.php?ids=108742,108755

### CODE REVIEW

**Comment by HBelusca** [comment]
Out of curiosity, has there been known applications that depend on this?

**Comment by xmine64** [comment]
> Out of curiosity, has there been known applications that depend on this?

As mentioned by simonelombardo in Jira and Mattermost, this allows loading original IE-based MSHTML (Instead of the WineGecko-based one that ReactOS ships)

File: modules/rostests/apitests/gdi32/CMakeLists.txt [hunk 1]
```diff
@@ -77,7 +77,8 @@ list(APPEND SOURCE
     SetWindowExtEx.c
     SetWorldTransform.c
     StretchBlt.c
-    TextTransform.c)
+    TextTransform.c
+    CreateHalftonePalette.c)
```
Comment (SergeGautherie) [inline modules/rostests/apitests/gdi32/CMakeLists.txt]: Keep it sorted.

File: win32ss/gdi/ntgdi/palette.c [hunk 2]
```diff
@@ -528,12 +528,12 @@ NtGdiCreateHalftonePalette(HDC  hDC)
...
     PALETTEENTRY PalEntries[256];
-    PPALETTE ppal;
-    PDC pdc;
+    PPALETTE ppal = NULL;
+    PDC pdc = NULL;
     HPALETTE hpal = NULL;
 
     pdc = DC_LockDc(hDC);
-    if (!pdc)
+    if (hDC && !pdc)
```
Comment (SergeGautherie) [inline win32ss/gdi/ntgdi/palette.c]: Or you could use `if (pdc)` for this whole block (to skip the `DC_LockDc()` call), in sync with what you do below.

Comment (xmine64) [reply win32ss/gdi/ntgdi/palette.c]: This makes the block nest one more level but removes unnecessary calls in cases where HDC is NULL. Probably it's better to skip the call to DC_LockDc when HDC is NULL.

Comment (xmine64) [reply win32ss/gdi/ntgdi/palette.c]: Simone said it was done this way to minimize the count of changed lines. I also found that it's possible to optimize the path further by keeping every DC-related thing inside a `if (hDC)` block. I'll do it and run GDI32 tests, and push it if it works.

**Comment by xmine64** [comment]
A small refactor has been done, and they probably won't break anything.

image

File: win32ss/gdi/ntgdi/palette.c [hunk 3]
```diff
@@ -553,13 +547,29 @@ NtGdiCreateHalftonePalette(HDC  hDC)
...
-        UNIMPLEMENTED;
+        pdc = DC_LockDc(hDC);
+        if (!pdc)
+        {
+            EngSetLastError(ERROR_INVALID_HANDLE);
+            return NULL;
+        }
+
+        ppal = PALETTE_ShareLockPalette(pdc->dclevel.hpal);
+        if (ppal && (ppal->flFlags & PAL_INDEXED))
```
Comment (SergeGautherie) [inline win32ss/gdi/ntgdi/palette.c]: Split this `if`, so `PALETTE_ShareUnlockPalette(ppal)` remains conditional.

File: win32ss/gdi/ntgdi/palette.c [hunk 4]
```diff
@@ -528,16 +528,10 @@ NtGdiCreateHalftonePalette(HDC  hDC)
 {
     int i, r, g, b;
     PALETTEENTRY PalEntries[256];
-    PPALETTE ppal;
-    PDC pdc;
+    PPALETTE ppal = NULL;
+    PDC pdc = NULL;
```
Comment (SergeGautherie) [inline win32ss/gdi/ntgdi/palette.c]: These changes are not needed anymore.

File: modules/rostests/apitests/gdi32/CreateHalftonePalette.c [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+ * COPYRIGHT:   Copyright 2026 Mohammad Amin Mollazadeh <madamin@pm.me>
+ */
+
+#include <apitest.h>
+#include <wingdi.h>
+
+START_TEST(CreateHalftonePalette)
+{
+    HPALETTE hPal;
+    UINT entries = 0;
```
Comment (SergeGautherie) [inline modules/rostests/apitests/gdi32/CreateHalftonePalette.c]: This init is superfluous.

File: modules/rostests/apitests/gdi32/CreateHalftonePalette.c [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+    entries = GetObjectW(hPal, sizeof(pe), pe);
+
+    ok(entries > 0, "Palette has no entries!\n");
+    ok(entries <= 256, "Unexpected number of palette entries: %u\n", entries);
+
+    /* basic sanity check: first entry should exist */
+    if (entries > 0)
+    {
+        ok(pe[0].peFlags <= 0xFF, "Invalid palette entry flags: %u\n", pe[0].peFlags);
```
Comment (SergeGautherie) [inline modules/rostests/apitests/gdi32/CreateHalftonePalette.c]: Initialize this (one) variable to an "invalid" value, earlier, as a double-check.

File: modules/rostests/apitests/gdi32/CreateHalftonePalette.c [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    if (!hPal)
+    {
+        skip("CreateHalftonePalette(NULL) failed, skipping test\n");
+        return;
+    }
+
+    entries = GetObjectW(hPal, sizeof(pe), pe);
+
+    ok(entries > 0, "Palette has no entries!\n");
+    ok(entries <= 256, "Unexpected number of palette entries: %u\n", entries);
```
Comment (SergeGautherie) [inline modules/rostests/apitests/gdi32/CreateHalftonePalette.c]: Use `#define ... 256`, or at least `_countof(pe)`?

File: modules/rostests/apitests/gdi32/CreateHalftonePalette.c [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+    /* basic sanity check: first entry should exist */
+    if (entries > 0)
+    {
+        ok(pe[0].peFlags <= 0xFF, "Invalid palette entry flags: %u\n", pe[0].peFlags);
+    }
+
+    ok(DeleteObject(hPal) == TRUE, "DeleteObject(hPal) failed\n");
```
Comment (SergeGautherie) [inline modules/rostests/apitests/gdi32/CreateHalftonePalette.c]: `== TRUE` looks unneeded/unwanted.

File: modules/rostests/apitests/gdi32/CreateHalftonePalette.c (hunk 6 above)
Comment (xmine64) [reply modules/rostests/apitests/gdi32/CreateHalftonePalette.c]: I tried doing this but the test failed on Windows 7, looks like GetObjectW on Windows 7 leaves it uninitialized.

Comment (xmine64) [reply modules/rostests/apitests/gdi32/CreateHalftonePalette.c]: My bad, I had to use GetPaletteEntries instead of GetObjectW.

**Comment by xmine64** [comment]
Updated:

image

File: modules/rostests/apitests/gdi32/CreateHalftonePalette.c [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+    /* First and last ten entries are default ones */
+    for (int i = 0; i < 10; i++)
+    {
+        ok(pe[i].peRed == sysPal[i].peRed && 
+            pe[i].peGreen == sysPal[i].peGreen &&
+            pe[i].peBlue == sysPal[i].peBlue, "Unexpected color at index %d\n", i);
+
+        ok(pe[246 + i].peRed == sysPal[10 + i].peRed && 
+            pe[246 + i].peGreen == sysPal[10 + i].peGreen &&
+            pe[246 + i].peBlue == sysPal[10 + i].peBlue, "Unexpected color at index %d\n", 246 + i);
```
Comment (HBelusca) [inline modules/rostests/apitests/gdi32/CreateHalftonePalette.c]:
```suggestion
           pe[i].peGreen == sysPal[i].peGreen &&
           pe[i].peBlue == sysPal[i].peBlue, "Unexpected color at index %d\n", i);

        ok(pe[246 + i].peRed == sysPal[10 + i].peRed && 
           pe[246 + i].peGreen == sysPal[10 + i].peGreen &&
           pe[246 + i].peBlue == sysPal[10 + i].peBlue, "Unexpected color at index %d\n", 246 + i);
```
(spacing nit)

File: modules/rostests/apitests/gdi32/CreateHalftonePalette.c [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+ * COPYRIGHT:   Copyright 2026 Mohammad Amin Mollazadeh <madamin@pm.me>
+ */
+
+#include <apitest.h>
+#include <wingdi.h>
+
+START_TEST(CreateHalftonePalette)
+{
+    HPALETTE hPal;
+    UINT entries;
```
Comment (HBelusca) [inline modules/rostests/apitests/gdi32/CreateHalftonePalette.c]:
```suggestion
    UINT entries;
    BOOL ret;
```

File: modules/rostests/apitests/gdi32/CreateHalftonePalette.c [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+        ok(pe[i].peRed == sysPal[i].peRed && 
+            pe[i].peGreen == sysPal[i].peGreen &&
+            pe[i].peBlue == sysPal[i].peBlue, "Unexpected color at index %d\n", i);
+
+        ok(pe[246 + i].peRed == sysPal[10 + i].peRed && 
+            pe[246 + i].peGreen == sysPal[10 + i].peGreen &&
+            pe[246 + i].peBlue == sysPal[10 + i].peBlue, "Unexpected color at index %d\n", 246 + i);
+    }
+
+    ok(DeleteObject(hPal), "DeleteObject(hPal) failed\n");
```
Comment (HBelusca) [inline modules/rostests/apitests/gdi32/CreateHalftonePalette.c]:
```suggestion
    ret = DeleteObject(hPal);
    ok(ret != FALSE, "DeleteObject(hPal) failed\n");
```

File: modules/rostests/apitests/gdi32/CreateHalftonePalette.c (hunk 9 above)
Comment (xmine64) [reply modules/rostests/apitests/gdi32/CreateHalftonePalette.c]: Sorry, in my own editor they looked like they need one more space, and I added all those manually to make them aligned vertically. Looks like the result wasn't the same as what it looked like in my editor.

**Review by HBelusca — APPROVED** [review]
LGTM, just running the testbots for the PR right now...

**Comment by HBelusca** [comment]
I'll merge it tomorrow by grouping the commits in two: first the test, second the fix.

**Comment by xmine64** [comment]
> I'll merge it tomorrow by grouping the commits in two: first the test, second the fix.

I think it would be better if you squash the commits. I didn't write proper commit messages.

**Comment by HBelusca** [comment]
@xmine64 Your PR has been merged!

Some things I had to change in your PR (and that you'll have to fix in your settings in the future):
- your author name in commits (I changed "Amin" to your full name, as was specified in the Copyright notice of the new apitest);
- you need to configure your git repository so that it automatically fix the ending newlines, with the `git config core.autocrlf true` (per-repo), or `git config --global core.autocrlf true` (globally). Because when rebasing your code I saw that some of the stuff in your new apitest file used only LF endlines, instead of CRLF (when on Windows). The git setting can take care of the conversion for text files automatically.
- You need to enable in your code editor, showing whitespace (blanks and tabs), so as to detect and remove/not add trailing whitespace, as I noticed was added in your new apitest file and the modified NtGdiCreateHalftonePalette code.

**Comment by xmine64** [comment]
Sorry, thanks for letting me know. I'm using a new environment for development and I forgot to configure some stuff.
